xwiki-platform-core/xwiki-platform-comment/xwiki-platform-comment-default/src/main/java/org/xwiki/platform/comment/internal/DefaultCommentManager.java

```diff
@@ -0,0 +1,92 @@
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.platform.comment.internal;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.EntityType;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.platform.comment.Comment;
+import org.xwiki.platform.comment.CommentException;
+import org.xwiki.platform.comment.CommentManager;
+import org.xwiki.user.UserReference;
+import org.xwiki.user.UserReferenceResolver;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+@Component
+@Singleton
+public class DefaultCommentManager implements CommentManager
+{
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private UserReferenceResolver<String> userReferenceResolver;
+
+    @Override
+    public List<Comment> getComments(EntityReference entityReference, boolean ascendindOrder)
+        throws CommentException
+    {
+        if (entityReference.getType() != EntityType.DOCUMENT) {
+            throw new CommentException("Current implementation only support document entities");
+        }
+        XWikiContext context = this.contextProvider.get();
+        try {
+            XWikiDocument document = context.getWiki().getDocument(entityReference, context);
+            List<BaseObject> comments = document.getComments();
+            List<Comment> result = new ArrayList<>();
+            for (BaseObject comment : comments) {
+                result.add(this.buildComment(comment));
+            }
+            Comparator<Comment> comparator = Comparator.comparing(Comment::getCreationDate);
+            if (!ascendindOrder) {
+                comparator = comparator.reversed();
+            }
+            result.sort(comparator);
+            return result;
+        } catch (XWikiException e) {
+            throw new CommentException(
+                String.format("Error when accessing comments for document [%s].", entityReference), e);
+        }
+    }
+
+    private DefaultComment buildComment(BaseObject commentObject) throws XWikiException
+    {
+        DefaultComment result = new DefaultComment(commentObject.getNumber());
```

Comment:
Hmm, wouldn't it make more sense for the comment id to really be unique (and not just unique for a given document) ?

Reply:
The commentId implementation depends on the storage. Right now on the storage the ID is unique for a given document. Whenever we'll move comments to not be xobject we'll be able to change that and this implem will allow it.

Reply:
The fact that the exact value depends on the implementation, sure. But the meaning of the id definitely cannot, and it needs to be clear in the API if the id is local to the entity or global.